libbindgen/tests/tests.rs

```diff
@@ -0,0 +1,150 @@
+extern crate diff;
+#[macro_use]
+extern crate lazy_static;
+extern crate libbindgen;
+extern crate tempdir;
+
+use std::env;
+use std::fs;
+use std::io::{BufReader, BufRead, Read};
+use std::path::{Path, PathBuf};
+use tempdir::TempDir;
+
+lazy_static! {
+    static ref TEMP_DIR: PathBuf = TempDir::new("libbindgen-tests")
+        .expect("Couldn't create temporary directory")
+        .into_path();
+}
+
+fn spawn_bindgen(header: &PathBuf, builder: libbindgen::Builder) -> Result<(), ()> {
+    let file_name = header.file_name()
+        .expect("spawn_bindgen expects a file");
+
+    let mut expected = PathBuf::from(header);
+    expected.pop();
+    expected.pop();
+    expected.push("expectations");
+    expected.push("tests");
+    expected.push(file_name);
+    expected.set_extension("rs");
+
+    let output = match builder.generate() {
+        Ok(bindings) => bindings.to_string(),
+        Err(_) => "".to_string(),
+    };
+
+    let mut buffer = String::new();
+    let _ = fs::File::open(&expected)
+        .expect("Couldn't read from expected test output")
+        .read_to_string(&mut buffer)
+        .expect("Couldn't read from expected test output");
+
+    if output == buffer {
+        Ok(())
+    } else {
+        println!("diff expected generated\n--- expected: {:?}\n+++ generated from: {:?}",
+            expected, header);
+        for diff in diff::lines(&buffer, &output) {
+            match diff {
+                diff::Result::Left(l) => println!("-{}", l),
+                diff::Result::Both(l, _) => println!(" {}", l),
+                diff::Result::Right(r) => println!("+{}", r),
+            }
+        }
+        Err(())
+    }
+}
+
+fn create_bindgen_builder(header: &PathBuf) -> libbindgen::Builder {
+    let mut builder = libbindgen::builder()
+        .header(header.to_str().unwrap())
+        .no_unstable_rust()
+        .raw_line("")
+        .raw_line("#![allow(non_snake_case)]")
+        .raw_line("");
+
+    let source = fs::File::open(header).unwrap();
+    let reader = BufReader::new(source);
+
+    let head: Result<Vec<_>, _> = reader.lines().take(5).collect();
+    let flagline = head.unwrap().into_iter()
+        .filter(|o| o.contains("bindgen-flags:")).nth(0);
+
+    if let Some(flagline) = flagline {
+        // FIXME: split flags again on =
+        let flags: Vec<_> = flagline.split("bindgen-flags:").last().unwrap()
+            .trim().split_whitespace().collect();
+        let mut it = flags.into_iter();
+        while let Some(flag) = it.next() {
+            builder = match flag {
+                "--enable-cxx-namespaces" => builder.enable_cxx_namespaces(),
```

Comment:
I was thinking that when I do the clap changes, I could modularise the parsing to use it here and in the binary.

`run-bindgen.py` still generates expectations (see `Makefile`), so that'll need some follow-up work.